AirCasting/SessionViews/ChartViewModel.swift

```diff
@@ -13,9 +13,17 @@ final class ChartViewModel: ObservableObject {
     @Published var chartEndTime: Date?
     @ObservedObject var session: SessionEntity
 
+    private var databaseObserver: ChartDatabaseObserver?
     var stream: MeasurementStreamEntity? {
         didSet {
-            guard session.isActive || session.isFollowed || session.status == .NEW else { return }
+            databaseObserver = nil
+            guard shouldBeUpdatingChartForCurrentSession(), let sensor = stream?.sensorName else { return }
+            let filterBy: ChartDatabaseObserverFilter = session.isFixed ? .hour : .minute
+            databaseObserver = ChartDatabaseObserver(session: session.uuid, sensor: sensor, timedFilter: filterBy) { [weak self] in
```

Comment:
Perhaps we could have one ChartDatabaseObserver per chart that would notify chartViewModel about new measurements for this session and call the generateEntries function, but the charViewModel would know for which stream it should generate entries?

Reply:
We could, but IMO putting it into the minute/hour enum and moving the now-generic filtering logic inside observer logic is a bit more neat and will keep low-level logic away from viewmodel

Reply:
I'm not sure I understand your answer and if I explained myself well so I'll try again. I meant that we could not pass the sensor information to the chartDatabaseObserver and have just one observer for the whole session since we get measurements in each stream at once. The hour/minute filter would still be relevant, as it depends on the session type. I think we don't need to filter measurements in the observer, as we are doing it only doing it now to check if there are more than 0 measurements for that stream.

Reply:
> "we get measurements in each stream at once"

this is the key part IMO here. By doing so we'll be implicitly depending on this behavior and there is a very important (but not very obvious) code quality issue with that:
This is not a set-in-stone, core type of behavior that we should rely on. It's more like a side effect of our (or backend) previous design choices. It might change due to some completely different task in completely different module in the future and this class will break all of the sudden. This is called **fragile code** and should be avoided. We do so by being very verbose and specific here. We're operating in a context of specific session and stream, let's design this interface to be conformant with that.

Reply:
In the future we can actually move this observer to be hidden behind a protocol and inject different implementations based on session:

`Resolver.resolve(ChartMeasurementsObserver.self, args: [sessionId, sensorName])`

This way we can decouple viewmodel from database and also remove the need for conditional `if fixed`